### PR TITLE
add Migration Step that add uriParam transforms

### PR DESCRIPTION
**Pull request by holiday-sunrise**
each uriparam will be set as variable.

### CODE REVIEW

**Comment by holiday-sunrise** [comment]
Hi Ana @afelisatti  may you review the PR. thx

**Comment by afelisatti** [comment]
Hi, @holiday-sunrise. I'll take a look later this week, I was not aware of this limitation when migrating APIkit applications. I'll review some scenarios and check this PR out. Thanks again for your contributions!

File: mule-migration-tool-library/src/test/java/com/mulesoft/tools/migration/library/mule/domain/DomainTest.java [hunk 1]
```diff
@@ -160,8 +160,14 @@ public void setUp() throws Exception {
     when(appModel.getNodes(any(String.class)))
         .thenAnswer(invocation -> getElementsFromDocuments((String) invocation.getArguments()[0]));
     when(appModel.getNode(any(String.class)))
-        .thenAnswer(invocation -> getElementsFromDocuments((String) invocation.getArguments()[0]).iterator()
-            .next());
+        .thenAnswer(invocation -> {
```
Comment (afelisatti) [inline mule-migration-tool-library/src/test/java/com/mulesoft/tools/migration/library/mule/domain/DomainTest.java:163]: Was this broken by your changes? It looks unrelated.

**Review by afelisatti — CHANGES_REQUESTED** [review]
Hi, @holiday-sunrise. Sorry for the delayed review and thanks again for sharing this with us. I added several general comments around it. Other than that, would it be possible to add a test for the scenario or share a simple app I can add to our tests? Thanks!

**Comment by holiday-sunrise** [comment]
Hi ana I will do it next week. I am currently Bussy with other work. Kind regards rainte

File: mule-migration-tool-library/src/test/java/com/mulesoft/tools/migration/library/mule/domain/DomainTest.java (hunk 1 above)
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/test/java/com/mulesoft/tools/migration/library/mule/domain/DomainTest.java:163]: Yes without the change i got a nullpointer

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/tasks/PostprocessMuleApplication.java [hunk 2]
```diff
@@ -49,7 +50,7 @@ public String getFrom() {
 
   @Override
   public List<MigrationStep> getSteps() {
-    return newArrayList(new AttributesToInboundPropertiesScriptGenerator());
+    return newArrayList(new AttributesToInboundPropertiesScriptGenerator(), new MigrateApiConfigFlows());
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/tasks/PostprocessMuleApplication.java:53]: it must be a post migration step because during the API Kit migration the ApplicationModel is not yet saved. Also the UriParam transformation must be the first MuleEvent in a flow for the metadata to work.

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    if (builder == null) {
+      final String ramlLocation = getApplicationModel().getPomModel().get().getArtifactId() + ".raml";
+      ramlFile =
+          getApplicationModel().getProjectBasePath().resolve(MULE_4_API_FOLDER + File.separator + ramlLocation).toFile();
+      builder = new RamlModelBuilder();
+      result = builder.buildApi(ramlFile);
+    }
+
+    if (result.getApiV10() == null) {
+      throw new RuntimeException("Ramlfile invalid: " + result.getValidationResults().toString());
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:125]: Thats a good idea, may you define the message ?!?

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+      }
+    }
+  }
+
+  private List<String> getResourceByFlowName(final String flowName) throws RuntimeException {
+    List<String> uriParams = null;
+    File ramlFile = null;
+    if (builder == null) {
+      final String ramlLocation = getApplicationModel().getPomModel().get().getArtifactId() + ".raml";
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:117]: no because it it not set during the migration.

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2020, Mulesoft, LLC. All rights reserved.
+ * Use of this source code is governed by a BSD 3-Clause License
+ * license that can be found in the LICENSE.txt file.
+ */
+package com.mulesoft.tools.migration.library.mule.steps.core;
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:6]: it must be a post migration step because during the API Kit migration the ApplicationModel is not yet saved. Also the UriParam transformation must be the first MuleEvent in a flow for the metadata to work.

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        for (String name : uriParams) {
+          final Element variableElement = new Element("set-variable", CORE_EE_NAMESPACE);
+          variableElement.setAttribute("variableName", name);
+          variableElement.setContent(new CDATA("attributes.uriParams." + name));
+          variablesElement.addContent(variableElement);
+        }
+
+
+        final List<Element> childrensCopy = new ArrayList<Element>();
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:90]: i will check this

**Comment by holiday-sunrise** [comment]
Add a Example API

[enhancement-526-api.zip](https://github.com/mulesoft/mule-migration-assistant/files/7072289/enhancement-526-api.zip)

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+      apiConfig = node.getAttribute("config-ref").getValue();
+
+    }
+    addNameSpace(CORE_EE_NAMESPACE, EE_NAMESPACE_SCHEMA, element.getDocument());
+
+
+
+    final String flowName = element.getAttributeValue("name");
+
+    if (flowName.endsWith(":api-config")) {
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:72]: it was a failure. changed

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:90]: yes this will work thanks for the hint

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (afelisatti) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:113]: Perhaps `API specification could not be analyzed: URI parameters variables will not be introduced in corresponding flows.`?

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/tasks/PostprocessMuleApplication.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (afelisatti) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/tasks/PostprocessMuleApplication.java:53]: That's fine. What I meant is that it should be named as a post processor specific to Apikit and its code should live with the other Apikit migration components. Then it can be added in the correct order in the method I referenced above.

File: mule-migration-tool-library/src/main/resources/report.yaml [hunk 8]
```diff
@@ -1070,6 +1070,9 @@ raml:
   noMatchingConfig:
     type: ERROR
     message: No matching config was found for RAML element.
+  invalid:
+    type: ERROR
+    message: "Ramlfile invalid: '{syntax}' Please, verify the raml syntax."
```
Comment (afelisatti) [inline mule-migration-tool-library/src/main/resources/report.yaml:1075]: Proposed message: `API specification could not be analyzed: URI parameters variables will not be introduced in corresponding flows.`

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (afelisatti) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:102]: What do you mean it's not set? I meant in this element:
`<apikit:config name="api-config" raml="api.raml" outboundHeadersMapName="outboundHeaders" httpStatusVarName="httpStatus" />`
You have the name of the API and don't have to infer it from the artifact ID.

**Review by holiday-sunrise — COMMENTED** [review]
fine for me, should we add the failure

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+          transformElement.removeContent();
+          final Element variablesElement = new Element("variables", CORE_EE_NAMESPACE);
+          variablesElement.removeContent();
+          transformElement.addContent(variablesElement);
+
+          for (String name : uriParams) {
+            final Element variableElement = new Element("set-variable", CORE_EE_NAMESPACE);
+            variableElement.setAttribute("variableName", name);
+            variableElement.setContent(new CDATA("attributes.uriParams." + name));
+            variablesElement.addContent(variableElement);
```
Comment (afelisatti) [inline mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:87]: One thing I checked reviewing the APIkit cases is that this transform could be non final. Meaning, users could actually replace the references to these variables with the new references in most cases. So I think these transforms should generate a report entry calling for a review.  Saying something like: `Mule 4 provides access to URI parameters via the "attributes.uriParams" map instead of variables. Please review usages of the {uriParam} variable and validate whether setting it is still required or it could be replaced with a reference to attributes.uriParams.{uriParam}.`

Comment (afelisatti) [inline mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:87]:
```suggestion
            variablesElement.addContent(variableElement);
            report.report("raml.uriParams", element, variableElement, name);
```

Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:87]: hi but mule generates for a new API the same transformation ?!?! . Why should we add the requested report entry.

Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:87]: In a new API, the transformation for the URI parameters is created at the scaffold of the Raml. Thus, this should also be done during the migration and this without a hint. This confuses the user what is correct.

Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:87]: In a new API, the transformation for the URI parameters is created at the scaffold of the Raml. Thus, this should also be done during the migration and this without a hint. This confuses the user what is correct.

**Comment by holiday-sunrise** [comment]
@afelisatti  hi, any news ?

**Comment by holiday-sunrise** [comment]
Anything to do ?

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+      migrationReport.report("raml.invalid", element, element,
+                             result.getValidationResults().toString());
+      throw new RuntimeException("Ramlfile invalid: " + result.getValidationResults().toString());
+    }
+
+
+    final List<Resource> resources = result.getApiV10().resources();
+    boolean found = false;
+    for (Resource resource : resources) {
```
Comment (afelisatti) [inline mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:119]: I need to review this logic further, I'm wondering if we really need to use the API spec or we can simply trust that the APIkit flow name correctly identifies the URI params. So, just analyzing the flow name and adding as params everything in between `{}` as those should be the params.

**Review by afelisatti — CHANGES_REQUESTED** [review]
Hi, @holiday-sunrise. Let me know if you'll have time to address this. Otherwise we can evaluate merging the PR as is and making the changes ourselves, so we don't hold this back any longer.

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        }
+      }
+    }
+  }
+
+  private List<String> getResourceByFlowName(final String flowName, MigrationReport migrationReport, Element element)
+      throws RuntimeException {
+    List<String> uriParams = null;
+    File ramlFile = null;
+    if (builder == null) {
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:101]: i mule 3 you cant have more than one API in a APP

File: mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    this.setAppliedTo(XPATH_SELECTOR);
+  }
+
+  @Override
+  public void execute(Element element, MigrationReport report) throws RuntimeException {
+
+    // resolve config-ref
+    String apiConfig = null;
+
+    final Element node = getApplicationModel().getNode(XPATH_SELECTOR_ROUTER);
```
[the remark this answers is not in the document]
Comment (holiday-sunrise) [reply mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/core/MigrateApiConfigFlows.java:61]: No it is only possible to have 1 apikit config. never seen anything different

**Comment by holiday-sunrise** [comment]
try to merg it as it is. thx

**Comment by holiday-sunrise** [comment]
hi @afelisatti cant fix "salesforce cla" bot i am in a new company and may old mail address was deleted. What can we do ?

**Comment by afelisatti** [comment]
Ok, @holiday-sunrise. I will try and merge this to a branch of this project to continue working on it before pushing it to the main branch. Thanks for the contribution! We had missed this bit of the APIkit scaffolding.

**Review by afelisatti — APPROVED** [review]
